Tests/BugsnagMetadataTest.m

```diff
@@ -0,0 +1,184 @@
+
+@interface BugsnagMetadataTest : XCTestCase <BugsnagMetadataDelegate>
+@property BOOL delegateCalled;
+@end
+
+@interface BugsnagMetadata ()
+@property(atomic, strong) NSMutableDictionary *dictionary;
+@end
+
+// MARK: - DummyClass
+
+@interface DummyClass : NSObject <NSCopying> // <NSCopying> allows it to be used as a dictionary key
+@property NSString *name;
+@end
+
+@implementation DummyClass
+@synthesize name;
+- (nonnull id)copyWithZone:(nullable NSZone *)zone {
+    DummyClass *copy = [DummyClass new];
+    copy.name = self.name;
+    return copy;
+}
+
+@end
+
+// MARK: - Tests
+
+@implementation BugsnagMetadataTest
+
+@synthesize delegateCalled;
+
+- (void)test_addAttribute_withName {
+    
+    // Creation
+    delegateCalled = NO;
+    BugsnagMetadata *metadata = [[BugsnagMetadata alloc] init];
+    XCTAssertNotNil(metadata);
+    XCTAssertFalse(delegateCalled, "Did not expect the delegate's metadataChanged: method to be called.");
+    XCTAssertNil([metadata delegate]);
+    
+    metadata.delegate = self;
+    XCTAssertNotNil([metadata delegate]);
+    XCTAssertFalse(delegateCalled, "Did not expect the delegate's metadataChanged: method to be called.");
+    
+    // Arbitrary tab name creates and returns itself
+    delegateCalled = NO;
+    NSMutableDictionary *tab = [metadata getTab:@"unknown"];
+    XCTAssertNotNil(tab);
+    XCTAssertEqual(tab.count, 0);
+    XCTAssertEqual([[metadata toDictionary] count], 1);
+    XCTAssertFalse(delegateCalled, "Expected the delegate's metadataChanged: method to be called.");
+    
+    // Check that the arbitrary named tab was set.
+    delegateCalled = NO;
+    [metadata addAttribute:@"foo" withValue:@"aValue" toTabWithName:@"SecondTab"];
+    XCTAssertEqual([[metadata toDictionary] count], 2);
+    XCTAssertTrue(delegateCalled, "Expected the delegate's metadataChanged: method to be called.");
+    
+    [metadata addAttribute:@"foo" withValue:@"aValue" toTabWithName:@"FirstTab"];
+    NSMutableDictionary *tab2 = [metadata getTab:@"FirstTab"];
+    XCTAssertNotNil(tab2);
+    XCTAssertEqual(tab2.count, 1);
+    
+    [metadata addAttribute:@"bar" withValue:@"anotherValue" toTabWithName:@"FirstTab"];
+    tab2 = [metadata getTab:@"FirstTab"];
+    XCTAssertEqual(tab2.count, 2);
+    
+    NSDictionary *dict = [metadata toDictionary];
+    XCTAssertEqual([dict count], 3);
+    
+    delegateCalled = NO;
+    [metadata clearTab:@"FirstTab"];
+    tab2 = [metadata getTab:@"FirstTab"];
+    XCTAssertEqual(tab2.count, 0);
+    XCTAssertTrue(delegateCalled, "Expected the delegate's metadataChanged: method to be called.");
+
+    // Adding invalid values should fail silently (and not add the value)
+    delegateCalled = NO;
+    [metadata addAttribute:@"bar" withValue:[DummyClass new] toTabWithName:@"FirstTab"];
+    tab2 = [metadata getTab:@"FirstTab"];
+    XCTAssertEqual(tab2.count, 0);
+    XCTAssertTrue(delegateCalled, "Expected the delegate's metadataChanged: method to be called.");
```

Comment:
If _all_ the values are invalid and nothing has changed in the metadata, then we could probably avoid calling the delegate here - particularly if it does a lot of work. I'm unsure if this is realistically going to happen often, so would be good to get your judgement on whether you think it's worth making that change.

Reply:
Agreed.  I'd noted that.  It's small so I've made the change, ensuring that any delegate calls occur at most once per operation, and outside @synchronised sections.  Two tests had to be adjusted to check that the delegate is NOT called.